src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -189,13 +189,45 @@ private NamedTypeSymbol ConstructAnonymousDelegateImplementationSymbol(Anonymous
             // If all parameter types and return type are valid type arguments, construct
             // the delegate type from a generic template. Otherwise, use a non-generic template.
             bool useUpdatedEscapeRules = Compilation.SourceModule.UseUpdatedEscapeRules;
-            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr))
+            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr, out var needsIndexedName))
             {
                 var fields = typeDescr.Fields;
                 Debug.Assert(fields.All(f => hasDefaultScope(useUpdatedEscapeRules, f)));
 
-                bool returnsVoid = fields.Last().Type.IsVoidType();
+                bool returnsVoid = fields[^1].Type.IsVoidType();
                 int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
+                for (int i = 0; i < nTypeArguments; i++)
+                {
+                    typeArgumentsBuilder.Add(fields[i].TypeWithAnnotations);
+                }
+                var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
+
+                // Delegate name cannot be fully determined by its signature (e.g., it has default parameter values).
+                if (needsIndexedName)
+                {
+                    Debug.Assert(nTypeArguments != 0);
+
+                    // Replace field types (where type parameters will be) by placeholders.
+                    var genericFieldTypes = IndexedTypeParameterSymbol.Take(nTypeArguments);
+                    if (returnsVoid)
+                    {
+                        var genericFieldTypeBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(fields.Length);
+                        genericFieldTypeBuilder.AddRange(genericFieldTypes);
+                        genericFieldTypeBuilder.Add(fields[^1].TypeWithAnnotations);
+                        genericFieldTypes = genericFieldTypeBuilder.ToImmutableAndFree();
+                    }
+                    var genericTypeDescr = typeDescr.WithNewFieldsTypes(genericFieldTypes);
```

Comment:
>var genericTypeDescr = typeDescr.WithNewFieldsTypes(genericFieldTypes); [](http://example.com/codeflow?start=20&length=71)

>> Consider separating blocks from other statements with an empty line. For the sake of readability

This is a general feedback. Consider always adding an empty line after a closing brace. For example, this is applicable to the code above as well:
```
                }
                var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
```

If you want to group a bunch of statements together, a block can be used for that, or it can be done with comments, which is going to improve readability even more.

Reply:
Thanks, changed.